cms/core/cache.py

```diff
@@ -36,3 +42,61 @@ def get_default_cache_control_decorator() -> Callable:
     """
     cache_control_kwargs = get_default_cache_control_kwargs()
     return cache_control(**cache_control_kwargs)
+
+
+class ElastiCacheIAMCredentialProvider(redis.CredentialProvider):
+    """A custom redis credential provider to use IAM for authentication.
+
+    https://redis.readthedocs.io/en/stable/examples/connection_examples.html#Connecting-to-a-redis-instance-with-ElastiCache-IAM-credential-provider.
+    """
+
+    # Authentication tokens are only valid for a maximum of 15 minutes.
+    TOKEN_TTL = 900
+
+    def __init__(self, user: str, cluster_name: str, region: str):
+        self.user = user
+        self.cluster_name = cluster_name
+        self.region = region
+
+        session = botocore.session.get_session()
+        self.request_signer = RequestSigner(
+            ServiceId("elasticache"),
+            self.region,
+            "elasticache",
+            "v4",
+            session.get_credentials(),
+            session.get_component("event_emitter"),
+        )
+
+        self.cache_key = f"elasticache_{user}_{cluster_name}_{region}"
+
+        self.connection_url = urlunparse(
```

Comment:
question: the redis example caches the connection URL under `get_credentials`. Is this used anywhere else? If not, why not keep it out of init?

Reply:
I moved it to the constructor so it's only generated once, since it doesn't change per call or based on any parameters.